support file-groups and improve review requests for maintainer file changes

nashif opened the pull request:
- **scripts: get_maintainer: support file groups**
- **scripts: set_assignee.py: Support file groups**
- **MAINTAINERS file: add documentation for file groups**
- **MAINTAINERS file: add description for tests**
- **ci: assigner: merge maintainer check into assigner workflow**
- **scripts: move set_assignee.py into scripts/ci**
- **scripts: set_assignee: request review from maintainers of changed areas**

nashif committed:
scripts: get_maintainer: support file groups

This new section allows defining a group of files in an area and makes
it possible to assign collaborators to the file group being defined.

The purpose of this new section is to allow fine tuning who is added as
reviewer when files change in a group. It is especially useful in large
areas with hundreds of files, for example platform areas.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
ci: assigner: merge maintainer check into assigner workflow

Merge two workflows into one for code sharing an efficiency.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
scripts: set_assignee: request review from maintainers of changed areas

Also request reviewes from maintainers of changes areas in the
maintainer file.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
scripts: move set_assignee.py into scripts/ci

Scripts only used by CI, so move it into that directory.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
scripts: set_assignee.py: Support file groups

Deal with new section in the maintainer file defining file groups.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
MAINTAINERS file: add documentation for file groups

Document file groups and how they should be used.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

aescolar commented on MAINTAINERS.yml:
```diff
@@ -60,6 +60,22 @@
+#           files:
+#             - <file path>
+#             - <file path>
+#           files-regex:
```
question: does it also support `files-exclude` and `files-regex-exclude`?

nashif replied:
added those now

aescolar commented on MAINTAINERS.yml:
```diff
@@ -32,6 +32,9 @@
 #         List of GitHub labels to add to pull requests that modify the area.
 #
+#     tests:
+#         List of test identifiers to be run when this area is affected by a PR.
```
Can you elaborate what do you mean with test identifiers?
Will all twister tests which identifier start with this prefix(es) be run?

aescolar reviewed (commented):
Thanks Anas, 2 questions

aescolar reviewed (commented):
+1 (I haven't tested it)

maass-hamburg commented on MAINTAINERS.yml:
```diff
@@ -60,6 +60,28 @@
+#         different parts of the area.
+#         Each group should have the following structure:
+#         - name: <group name>
+#           collaborators:
```
these collaborators are added in addition to the area collaborators, right? Could maybe be mentioned.

nashif committed:
scripts: get_maintainer: file group pattern inherit top area patterns

File groups inherit file patterns from their parent area. A file will only
match a file group if it first matches the parent area's patterns, and then
also matches the file group's own patterns. This allows file groups to
further filter and subdivide files that are already covered by the area.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

decsny commented:
It would be nice if the get_maintainer script commands had some more intelligence about file groups, like if the `list` command worked with file group names and if the `path` command only listed what file groups are relevant under a matched area instead of all the file groups

nashif commented on MAINTAINERS.yml:
```diff
@@ -60,6 +60,34 @@
+#         File groups inherit file patterns from their parent area. A file will only
+#    			match a file group if it first matches the parent area's patterns, and then
+#    			also matches the file group's own patterns. This allows file groups to
+#    			further filter and subdivide files that are already covered by the area.#
```
fixed

fabiobaltieri commented on .github/workflows/assigner.yml:
```diff
@@ -71,4 +72,13 @@ jobs:
+      env:
+        GITHUB_TOKEN: ${{ secrets.ZB_PR_ASSIGNER_GITHUB_TOKEN }}
+      run: |
+        python  ./scripts/ci/check_maintainer_changes.py  \
```
I would have added it but you asked to only do it on issues...

aescolar reviewed (approved):
+1
